x/protorev/keeper/rebalance.go

```diff
@@ -167,30 +173,151 @@ func (k Keeper) FindMaxProfitForRoute(ctx sdk.Context, route RouteMetaData, rema
+	inputDenom string,
+	inputMap map[uint64]string,
+	curLeft, curRight sdk.Int,
+) (sdk.Int, error) {
+	// Find the max amount in that can be used for the binary search (in terms of uosmo)
+	maxInputAmount, stepSize, err := k.MaxInputAmount(ctx)
+	if err != nil {
+		return sdk.ZeroInt(), err
+	}
+
+	// Iterate through all CL pools and determine the maximal amount of input that can be used
+	// respecting the max ticks moved.
+	for _, poolId := range clPools {
+		maxInForPool, _, err := k.concentratedLiquidityKeeper.ComputeMaxInAmtGivenMaxTicksCrossed(
+			ctx,
+			poolId,
+			inputMap[poolId],
+			types.MaxTicksMoved,
+		)
+
+		// In the case where we cannot calculate the max in amount, we short circuit the search
+		// and return an upper bound of 0. This will cause the binary search to not run.
+		if err != nil {
+			return sdk.ZeroInt(), err
+		}
+
+		maxInForPoolInOsmo, err := k.ConvertProfits(ctx, maxInForPool, maxInForPool.Amount)
+		if err != nil {
+			return sdk.ZeroInt(), err
+		}
+
+		if maxInForPoolInOsmo.LT(maxInputAmount) {
+			maxInputAmount = maxInForPoolInOsmo
+		}
+	}
+
+	return maxInputAmount.Quo(stepSize), nil
+}
+
 // Determine if the binary search range needs to be extended
-func (k Keeper) ExtendSearchRangeIfNeeded(ctx sdk.Context, route RouteMetaData, inputDenom string, curLeft, curRight sdk.Int) (sdk.Int, sdk.Int) {
+func (k Keeper) ExtendSearchRangeIfNeeded(
+	ctx sdk.Context,
+	route RouteMetaData,
+	inputDenom string,
+	curLeft, curRight, updatedMax sdk.Int,
+) (sdk.Int, sdk.Int, error) {
 	// Get the profit for the maximum amount in
 	_, maxInProfit, err := k.EstimateMultihopProfit(ctx, inputDenom, curRight.Mul(route.StepSize), route.Route)
 	if err != nil {
-		return curLeft, curRight
+		return sdk.ZeroInt(), sdk.ZeroInt(), err
 	}
 
 	// If the profit for the maximum amount in is still increasing, then we can increase the range of the binary search
 	if maxInProfit.GTE(sdk.ZeroInt()) {
+		maxInPlusOne := curRight.Add(sdk.OneInt()).Mul(route.StepSize)
+		if updatedMax.LT(maxInPlusOne) {
+			return curLeft, curRight, nil
+		}
+
 		// Get the profit for the maximum amount in + 1
-		_, maxInProfitPlusOne, err := k.EstimateMultihopProfit(ctx, inputDenom, curRight.Add(sdk.OneInt()).Mul(route.StepSize), route.Route)
+		_, maxInProfitPlusOne, err := k.EstimateMultihopProfit(ctx, inputDenom, maxInPlusOne, route.Route)
 		if err != nil {
-			return curLeft, curRight
+			return sdk.ZeroInt(), sdk.ZeroInt(), err
 		}
 
 		// Change the range of the binary search if the profit is still increasing
 		if maxInProfitPlusOne.GT(maxInProfit) {
 			curLeft = curRight
-			curRight = types.ExtendedMaxInputAmount
+			curRight = updatedMax
 		}
 	}
 
-	return curLeft, curRight
+	return curLeft, curRight, nil
+}
+
+// MaxInputAmount returns the max input amount that can be used in any route. We use uosmo as the base
+// unit of account for the max input amount.
+func (k Keeper) MaxInputAmount(ctx sdk.Context) (sdk.Int, sdk.Int, error) {
```

Comment:
Re: our in-person sync, if we go the reverse iteration method I don't think we need to choose any base unit of account for this so should be able to remove